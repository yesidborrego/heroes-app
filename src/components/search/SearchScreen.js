import React, { useMemo } from 'react'
import queryString from "query-string";
import { useLocation } from 'react-router-dom';
import { useForm } from '../../hooks/useForm';
import { HeroCard } from '../heroes/HeroCard';
import { getHeroesBySearch } from '../../selectors/getHeroesBySearch';

const SearchScreen = ({ history }) => {
  const location = useLocation();
  const { q = ''} = queryString.parse(location.search);

  const [{searchHero}, handleInputChange] = useForm({
    searchHero: q
  });

  const heroesFiltered = useMemo(() => getHeroesBySearch(q), [q]);

  const handleSearch = (e) => {
    e.preventDefault();
    history.push(`?q=${searchHero}`);
  };

  return (
    <div className="animate__animated animate__fadeIn">
      <h1>Search Screen</h1>
      <hr />
      <div className="row">
        <div className="col-5">
          <h4>Serach Form</h4>
          <hr />
          <form onSubmit={handleSearch}>
            <input
              type="text"
              name="searchHero"
              placeholder="Find your hero"
              className="form-control"
              onChange={handleInputChange}
              value={searchHero}
            />
            <div className="d-grid">
              <button
                type="submit"
                className="btn btn-primary mt-3 btn-block"
              >
                Search
              </button>
            </div>
          </form>
        </div>

        <div className="col-7">
          <h4>Results</h4>
          <hr />
          {
            (q === '')
            ?
              <div className="alert alert-info alert alert-info animate__animated animate__fadeIn">
                Search a Hero
              </div>
            :
              (heroesFiltered.length === 0 ) &&
              <div className="alert alert-danger animate__animated animate__zoomIn">
                Heroes not found
              </div>
          }
          {
            heroesFiltered.map(hero => (
              <HeroCard
                key={hero.id}
                hero={hero}
              />
            ))
          }
        </div>
      </div>
    </div>
  )
}

export default SearchScreen;
