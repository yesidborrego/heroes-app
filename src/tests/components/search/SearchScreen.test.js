import '@testing-library/jest-dom';
import { mount } from 'enzyme';
import { MemoryRouter, Route } from 'react-router-dom';
import SearchScreen from '../../../components/search/SearchScreen';

describe('Probar el componente <SearchScreen/>', () => {
  test('Debe mostrarse correctamente el componente', () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={['/search']}>
        <Route path="/search" component={SearchScreen} />
      </MemoryRouter>
    );

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('.alert-info').text().trim()).toBe('Search a Hero');
  })

  test('Debe cargar los datos de "Batman" y mostarlo en el input', () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={['/search?q=batman']}>
        <Route path="/search" component={SearchScreen} />
      </MemoryRouter>
    );

    expect(wrapper.find('input').prop('value')).toBe('batman');
    expect(wrapper.find('HeroCard').exists()).toBe(true);
    expect(wrapper).toMatchSnapshot();
  })

  test('Debe cargar el "alert-danger" si no se encuentra el hero', () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={['/search?q=batman1']}>
        <Route path="/search" component={SearchScreen} />
      </MemoryRouter>
    );
    expect(wrapper.find('.alert-danger').exists()).toBe(true);
    expect(wrapper.find('.alert-danger').text()).toBe('Heroes not found');
    expect(wrapper.find('HeroCard').exists()).toBe(false);
    expect(wrapper).toMatchSnapshot();
  })

  test('Debe llamar el "push" del history', () => {
    const historyMock = {
      push: jest.fn()
    };

    const wrapper = mount(
      <MemoryRouter initialEntries={['/search?q=batman']}>
        <Route
          path="/search"
          component={() => <SearchScreen history={historyMock} />} />
      </MemoryRouter>
    );
    wrapper.find('input').simulate('change', {
      target: {
        name: 'searchHero',
        value: 'batman'
      }
    });
    wrapper.find('form').prop('onSubmit')({
      preventDefault(){}
    });
    expect(historyMock.push).toHaveBeenCalledWith('?q=batman');
  })

})
