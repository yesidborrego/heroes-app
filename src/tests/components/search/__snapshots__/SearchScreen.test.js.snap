// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Probar el componente <SearchScreen/> Debe cargar el "alert-danger" si no se encuentra el hero 1`] = `
<div
  className="animate__animated animate__fadeIn"
>
  <h1>
    Search Screen
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Serach Form
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          className="form-control"
          name="searchHero"
          onChange={[Function]}
          placeholder="Find your hero"
          type="text"
          value="batman1"
        />
        <div
          className="d-grid"
        >
          <button
            className="btn btn-primary mt-3 btn-block"
            type="submit"
          >
            Search
          </button>
        </div>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="alert alert-danger animate__animated animate__zoomIn"
      >
        Heroes not found
      </div>
    </div>
  </div>
</div>
`;

exports[`Probar el componente <SearchScreen/> Debe cargar los datos de "Batman" y mostarlo en el input 1`] = `
<div
  className="animate__animated animate__fadeIn"
>
  <h1>
    Search Screen
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Serach Form
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          className="form-control"
          name="searchHero"
          onChange={[Function]}
          placeholder="Find your hero"
          type="text"
          value="batman"
        />
        <div
          className="d-grid"
        >
          <button
            className="btn btn-primary mt-3 btn-block"
            type="submit"
          >
            Search
          </button>
        </div>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="col animate__animated animate__fadeIn"
      >
        <div
          className="card m-3 border-warning"
          style={
            Object {
              "maxWidth": 540,
            }
          }
        >
          <div
            className="row g-0"
          >
            <div
              className="col-md-4"
            >
              <img
                alt="Batman"
                className="card-img"
                src="./assets/heroes/dc-batman.jpg"
              />
            </div>
            <div
              className="col-md-8"
            >
              <div
                className="card-body"
              >
                <h5
                  className="card-text"
                >
                  Batman
                </h5>
                <a
                  href="/hero/dc-batman"
                  onClick={[Function]}
                >
                  See more...
                </a>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Probar el componente <SearchScreen/> Debe mostrarse correctamente el componente 1`] = `
<div
  className="animate__animated animate__fadeIn"
>
  <h1>
    Search Screen
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Serach Form
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          className="form-control"
          name="searchHero"
          onChange={[Function]}
          placeholder="Find your hero"
          type="text"
          value=""
        />
        <div
          className="d-grid"
        >
          <button
            className="btn btn-primary mt-3 btn-block"
            type="submit"
          >
            Search
          </button>
        </div>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="alert alert-info alert alert-info animate__animated animate__fadeIn"
      >
        Search a Hero
      </div>
    </div>
  </div>
</div>
`;
